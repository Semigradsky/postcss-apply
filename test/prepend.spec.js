import postcss from 'postcss';
import { stripIndent } from 'common-tags';
import plugin from '../src';

describe('prepend', () => {
  describe('should prepend sets from options', () => {
    it('from an object set', async () => {
      const input = stripIndent`
        .dummy {
          @apply --justatest;
        }
      `;

      const sets = {
        justatest: {
          padding: '0 1rem',
          fontSize: '1.4rem',
          color: 'tomato',
        },
      };

      const result = await postcss()
        .use(plugin({ sets }))
        .process(input, { from: undefined });

      expect(result.css).toMatchSnapshot();
    });

    it('from a string set', async () => {
      const input = stripIndent`
        .dummy {
          @apply --justatest;
        }
      `;

      const sets = {
        justatest: `
          fontSize: 1.4rem;

          @media (width >= 500px) {
            fontSize: 2.4rem;
          }
        `,
      };

      const result = await postcss()
        .use(plugin({ sets }))
        .process(input, { from: undefined });

      expect(result.css).toMatchSnapshot();
    });

    it('throws if the set is not an object or a string', async () => {
      const input = stripIndent`
        .dummy {
          @apply --justatest;
        }
      `;

      const sets = {
        justatest: () => {},
      };

      expect(() => {
        postcss() // eslint-disable-line no-unused-expressions
          .use(plugin({ sets }))
          .process(input, { from: undefined }).css;
      }).toThrow(
        'Unrecognized set type `function`, must be an object or string.'
      );
    });
  });

  it('should override sets from options with CSS declared ones', async () => {
    const input = stripIndent`
      :root {
        --justatest: {
          padding: none;
          color: orangeRed;
        }
      }

      .dummy {
        @apply --justatest;
      }
    `;

    const sets = {
      justatest: {
        padding: '0 1rem',
        fontSize: '1.4rem',
        color: 'tomato',
      },
    };

    const result = await postcss()
      .use(plugin({ sets }))
      .process(input, { from: undefined });

    expect(result.css).toMatchSnapshot();
  });

  it('should be able to be nested inside a CSS declared set', async () => {
    const input = stripIndent`
      :root {
        --from-css: {
          @apply --from-js;
        }
      }

      .dummy {
        @apply --from-css;
      }
    `;

    const sets = {
      'from-js': {
        color: 'tomato',
      },
    };

    const result = await postcss()
      .use(plugin({ sets }))
      .process(input, { from: undefined });

    expect(result.css).toMatchSnapshot();
  });
});
