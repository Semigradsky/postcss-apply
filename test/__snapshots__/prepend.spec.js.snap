// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepend should be able to be nested inside a CSS declared set 1`] = `
".dummy {
  color: tomato;
}"
`;

exports[`prepend should override sets from options with CSS declared ones 1`] = `
".dummy {
  padding: none;
  color: orangeRed;
}"
`;

exports[`prepend should prepend sets from options from a string set 1`] = `
".dummy {
  fontSize: 1.4rem;
          @media (width >= 500px) {
  fontSize: 2.4rem;
          }
}"
`;

exports[`prepend should prepend sets from options from an object set 1`] = `
".dummy {
  color: tomato;
  font-size: 1.4rem;
  padding: 0 1rem;
}"
`;
